python_modules/dagster/dagster/_core/execution/retries.py

```diff
@@ -69,3 +80,66 @@ def mark_attempt(self, key: str) -> None:
 
     def snapshot_attempts(self) -> Mapping[str, int]:
         return dict(self._attempts)
+
+
+def auto_reexecution_should_retry_run(instance: "DagsterInstance", run: DagsterRun):
+    """Determines if a run will be retried by the automatic reexcution system.
+    A run will retry if:
+    - it is failed.
+    - the number of max allowed retries is > 0 (max retries can be set via system setting or run tag).
+    - there have not already been >= max_retries retries for the run.
+
+    If the run failure reason was a step failure and the retry_on_asset_or_op_failure tag/system setting is set to false,
+    a warning message will be logged and the run will not be retried.
+
+    We determine how many retries have been launched for the run by looking at the size of the run group
+    (the set of runs that have the same root_run_id and the run with root_run_id). Since manually launched retries are
+    part of the run group, this means that if a user launches a manual retry of run A and then this function
+    is called because a retry for run A launched by the auto-reexecution system failed, the manual retry will be
+    counted toward max_retries.
+
+    There is potential that one "extra" retry will be launched by the automatic reexecution system
+    since manual retries could be happening in parallel with automatic retries. Here is
+    an illustrative example:
+    - Max retries is 3
+    - Run A fails
+    - The automatic reexecution system launches a retry of run A (A_1), which fails
+    - The automatic reexecution system launches a retry run A_1 (A_2), which fails
+    - This function is executing and has fetched the run_group for run A_2: (A, A_1, A_2)
+    - A user launches a manual retry of run A (A_m). The run group is now (A, A_1, A_2, A_m), but this function does
+    not have the updated run group
+    - Since the run group we've fetched is (A, A_1, A_2), this function will mark A_2 as `will_retry=true` and
+    run `A_3` will be launched. This is the "extra" retry, since usually manual retries are counted toward max_retries, but
+    in this case it was not.
+    """
+    from dagster._core.events import RunFailureReason
+
+    if run.status == DagsterRunStatus.FAILURE:
```

Comment:
could also do

if run.status != DagsterRunStatus.Failure:
  return False

for less indentation